# Reinforcement learning for tic-tac-toe

In which we solve the problem of teaching a neural network to play tic-tac-toe.

* First, we import libraries for this example:

In [1]:
import random
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.backend import reshape
from keras.utils.np_utils import to_categorical

Using TensorFlow backend.


* Next we write helper methods to initialize the board, get a list of valid moves, declare a winner, and print the current board.

In [2]:
# Get an empty board
#
# 0 indicates an empty space, 1 indicates an 'X' (player 1), and 2 indicates an 'O' (player 2)
#
# Initially the board is empty, so we return a 3x3 array of zeroes.
def initBoard():
    board = [
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]
    ]
    return board

# Print the current state of the board
def printBoard(board):
    for i in range(len(board)):
        for j in range(len(board[i])):
            mark = ' '
            if board[i][j] == 1:
                mark = 'X'
            elif board[i][j] == 2:
                mark = 'O'
            if (j == len(board[i]) - 1):
                print(mark)
            else:
                print(str(mark) + "|", end='')
        if (i < len(board) - 1):
            print("-----")

# Get a list of valid moves (indices into the board)
def getMoves(board):
    moves = []
    for i in range(len(board)):
        for j in range(len(board[i])):
            if board[i][j] == 0:
                moves.append((i, j))
    return moves

# Declare a winner
#
# -1 = game not over
#  0 = draw
#  1 = 'X' wins (player 1)
#  2 = 'O' wins (player 2)
def getWinner(board):
    candidate = 0
    won = 0
    
    # Check rows
    for i in range(len(board)):
        candidate = 0
        for j in range(len(board[i])):
            
            # Make sure there are no gaps
            if board[i][j] == 0:
                break
            
            # Identify the front-runner
            if candidate == 0:
                candidate = board[i][j]
            
            # Determine whether the front-runner has all the slots
            if candidate != board[i][j]:
                break
            elif j == len(board[i]) - 1:
                won = candidate
    
    if won > 0:
        return won
    
    # Check columns
    for j in range(len(board[0])):
        candidate = 0
        for i in range(len(board)):
            
            # Make sure there are no gaps
            if board[i][j] == 0:
                break
            
            # Identify the front-runner
            if candidate == 0:
                candidate = board[i][j]
            
            # Determine whether the front-runner has all the slots
            if candidate != board[i][j]:
                break
            elif i == len(board) - 1:
                won = candidate
    
    if won > 0:
        return won
    
    # Check diagonals
    candidate = 0
    for i in range(len(board)):
        if board[i][i] == 0:
            break
        if candidate == 0:
            candidate = board[i][i]
        if candidate != board[i][i]:
            break
        elif i == len(board) - 1:
            won = candidate
    
    if won > 0:
        return won
    
    candidate = 0
    for i in range(len(board)):
        if board[i][2 - i] == 0:
            break
        if candidate == 0:
            candidate = board[i][2 - i]
        if candidate != board[i][2 - i]:
            break
        elif i == len(board) - 1:
            won = candidate
    
    if won > 0:
        return won
    
    # Still no winner?
    if (len(getMoves(board)) == 0):
        # It's a draw
        return 0
    else:
        # Still more moves to make
        return -1

* Next, we test the helper methods to demonstrate that they work.

In [3]:
# Test helper methods
b = initBoard()
printBoard(b)
print(getWinner(b))
print(getMoves(b))

b[0][0] = 1
b[1][1] = 1
b[2][2] = 1
printBoard(b)
print(getWinner(b))

b[0][2] = 2
b[1][2] = 2
b[2][2] = 2
printBoard(b)
print(getWinner(b))

b[0][1] = 1
b[1][0] = 2
b[2][0] = 1
b[2][1] = 2
b[2][2] = 1
b[0][0] = 2
printBoard(b)
print(getWinner(b))

 | | 
-----
 | | 
-----
 | | 
-1
[(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
X| | 
-----
 |X| 
-----
 | |X
1
X| |O
-----
 |X|O
-----
 | |O
2
O|X|O
-----
O|X|O
-----
X|O|X
0


* Next, we take everything we've done so far and create a random game simulator

In [12]:
random.seed()

# Get best next move for the given player at the given board position
def bestMove(board, model, player, rnd=0):
    scores = []
    moves = getMoves(board)
    
    # Make predictions for each possible move
    for i in range(len(moves)):
        future = np.array(board)
        future[moves[i][0]][moves[i][1]] = player
        prediction = model.predict(future.reshape((-1, 9)))[0]
        if player == 1:
            winPrediction = prediction[1]
            lossPrediction = prediction[2]
        else:
            winPrediction = prediction[2]
            lossPrediction = prediction[1]
        drawPrediction = prediction[0]
        if winPrediction - lossPrediction > 0:
            scores.append(winPrediction - lossPrediction)
        else:
            scores.append(drawPrediction - lossPrediction)

    # Choose the best move with a random factor
    bestMoves = np.flip(np.argsort(scores))
    for i in range(len(bestMoves)):
        if random.random() * rnd < 0.5:
            return moves[bestMoves[i]]

    # Choose a move completely at random
    return moves[random.randint(0, len(moves) - 1)]

# Simulate a game
def simulateGame(p1=None, p2=None, rnd=0):
    history = []
    board = initBoard()
    playerToMove = 1
    
    while getWinner(board) == -1:
        
        # Chose a move (random or use a player model if provided)
        move = None
        if playerToMove == 1 and p1 != None:
            move = bestMove(board, p1, playerToMove, rnd)
        elif playerToMove == 2 and p2 != None:
            move = bestMove(board, p2, playerToMove, rnd)
        else:
            moves = getMoves(board)
            move = moves[random.randint(0, len(moves) - 1)]
        
        # Make the move
        board[move[0]][move[1]] = playerToMove
        
        # Add the move to the history
        history.append((playerToMove, move))
        
        # Switch the active player
        playerToMove = 1 if playerToMove == 2 else 2
        
    return history

# Simulate a game
history = simulateGame()
print(history)

[(1, (2, 1)), (2, (0, 1)), (1, (1, 0)), (2, (2, 2)), (1, (2, 0)), (2, (1, 2)), (1, (0, 2)), (2, (1, 1)), (1, (0, 0))]


In [5]:
# Reconstruct the board from the move list
def movesToBoard(moves):
    board = initBoard()
    for move in moves:
        player = move[0]
        coords = move[1]
        board[coords[0]][coords[1]] = player
    return board

board = movesToBoard(history)
printBoard(board)
print(getWinner(board))

X|O|X
-----
X|O|O
-----
O|X|X
0


* Next, we generate a set of simulated games to train our neural network and calculate some win statistics for each random player. We predict that 'X' (player 1) should have a slight edge due to first mover advantage, but that both players have roughly the same number of wins, draws, and losses overall.

In [11]:
games = [simulateGame() for _ in range(10000)]

In [7]:
# Aggregate win/loss/draw stats for a player
def gameStats(games, player=1):
    stats = {"win": 0, "loss": 0, "draw": 0}
    for game in games:
        result = getWinner(movesToBoard(game))
        if result == -1:
            continue
        elif result == player:
            stats["win"] += 1
        elif result == 0:
            stats["draw"] += 1
        else:
            stats["loss"] += 1
    
    winPct = stats["win"] / len(games) * 100
    lossPct = stats["loss"] / len(games) * 100
    drawPct = stats["draw"] / len(games) * 100

    print("Results for player %d:" % (player))
    print("Wins: %d (%.1f%%)" % (stats["win"], winPct))
    print("Loss: %d (%.1f%%)" % (stats["loss"], lossPct))
    print("Draw: %d (%.1f%%)" % (stats["draw"], drawPct))

gameStats(games)
print()
gameStats(games, player=2)

Results for player 1:
Wins: 51934 (51.9%)
Loss: 24212 (24.2%)
Draw: 23854 (23.9%)

Results for player 2:
Wins: 24212 (24.2%)
Loss: 51934 (51.9%)
Draw: 23854 (23.9%)


* As shown above, when moves are chosen at random, player 1 has a significant advantage! However, we know that when people play tic-tac-toe, it's easy to force a draw, so we'd expect similar behavior from a neural network trained to play tic-tac-toe.
* We choose a DNN architecture since we effectively want to predict the outcome of a game based on the given board state.
* The input for each cell is the board state, which we reshape into a flat array of 9 elements, each element of which can be a 0 (empty cell), 1 (player 1 move), or 2 (player 2 move).
* The output is the result of the game (win, loss, or draw). We use a one-hot encoded array for this.

In [13]:
def getModel():
    numCells = 9 # How many cells in a 3x3 tic-tac-toe board?
    outcomes = 3 # How many outcomes are there in a game? (draw, X-wins, O-wins)
    model = Sequential()
    model.add(Dense(200, activation='relu', input_shape=(9, )))
    model.add(Dropout(0.2))
    model.add(Dense(125, activation='relu'))
    model.add(Dense(75, activation='relu'))
    model.add(Dropout(0.1))
    model.add(Dense(25, activation='relu'))
    model.add(Dense(outcomes, activation='softmax'))
    model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['acc'], verbose=0)
    return model
model = getModel()
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 200)               2000      
_________________________________________________________________
dropout_3 (Dropout)          (None, 200)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 125)               25125     
_________________________________________________________________
dense_8 (Dense)              (None, 75)                9450      
_________________________________________________________________
dropout_4 (Dropout)          (None, 75)                0         
_________________________________________________________________
dense_9 (Dense)              (None, 25)                1900      
_________________________________________________________________
dense_10 (Dense)             (None, 3)                 78        
Total para

* Next, we reshape our data in preparation for training. 

In [15]:
# Get a set of board states labelled by who eventually won that game
def gamesToWinLossData(games):
    X = []
    y = []
    for game in games:
        winner = getWinner(movesToBoard(game))
        for move in range(len(game)):
            X.append(movesToBoard(game[:(move + 1)]))
            y.append(winner)

    X = np.array(X).reshape((-1, 9))
    y = to_categorical(y)
    
    # Return an appropriate train/test split
    trainNum = int(len(X) * 0.8)
    return (X[:trainNum], X[trainNum:], y[:trainNum], y[trainNum:])

# Split out train and validation data
X_train, X_test, y_train, y_test = gamesToWinLossData(games)

* Next, we train our model on the random tic-tac-toe games we generated earlier.

In [16]:
nEpochs = 100
batchSize = 100
history = model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=nEpochs, batch_size=batchSize)

Train on 62306 samples, validate on 15577 samples
Epoch 1/100
62306/62306 [==============================] - 3s 41us/step - loss: 0.8223 - acc: 0.6135 - val_loss: 0.8075 - val_acc: 0.6164
Epoch 2/100
62306/62306 [==============================] - 2s 25us/step - loss: 0.8130 - acc: 0.6185 - val_loss: 0.7963 - val_acc: 0.6221
Epoch 3/100
62306/62306 [==============================] - 2s 29us/step - loss: 0.8056 - acc: 0.6214 - val_loss: 0.8031 - val_acc: 0.6237
Epoch 4/100
62306/62306 [==============================] - 2s 37us/step - loss: 0.8005 - acc: 0.6232 - val_loss: 0.7932 - val_acc: 0.6248
Epoch 5/100
62306/62306 [==============================] - 2s 38us/step - loss: 0.7976 - acc: 0.6247 - val_loss: 0.7943 - val_acc: 0.6257
Epoch 6/100
62306/62306 [==============================] - 2s 38us/step - loss: 0.7933 - acc: 0.6255 - val_loss: 0.7888 - val_acc: 0.6290
Epoch 7/100
62306/62306 [==============================] - 2s 38us/step - loss: 0.7907 - acc: 0.6272 - val_loss: 0.7824 - 

* We see that our very simple model caps out at around 58% validation accuracy.
* Next, we use this trained model to play a game and see if we do any better than random chance.

In [17]:
games2 = [simulateGame(p1=model) for _ in range(1000)]

In [18]:
gameStats(games2)

Results for player 1:
Wins: 950 (95.0%)
Loss: 16 (1.6%)
Draw: 34 (3.4%)


* Wow, we've gone from losing 24.2% of games to only losing 1.6% of the time! Let's see how our model fares as player 2.

In [19]:
games3 = [simulateGame(p2=model) for _ in range(1000)]

In [20]:
gameStats(games3, player=2)

Results for player 2:
Wins: 427 (42.7%)
Loss: 43 (4.3%)
Draw: 530 (53.0%)


* This is equally impressive! Player 2 has gone from losing 51.9% of games to only losing 4.3% of the time.
* Now what happens when the model is used to play both sides (with some random factor applied to force the model to choose different moves each game)?

In [22]:
games4 = [simulateGame(p1=model, p2=model, rnd=0.6) for _ in range(1000)]

In [23]:
gameStats(games4, player=1)
print()
gameStats(games4, player=2)

Results for player 1:
Wins: 483 (48.3%)
Loss: 44 (4.4%)
Draw: 473 (47.3%)

Results for player 2:
Wins: 44 (4.4%)
Loss: 483 (48.3%)
Draw: 473 (47.3%)


* We see that the game tends toward a draw, which is what we would expect from a couple of human players. Still, neither player is particularly good.

* As a final measure of skill, let's see how the length of the average game has changed. We would expect this to go down with an increased imbalance in skill levels.

In [24]:
print("Average length of fully random game is %f moves" % (np.mean([float(len(game)) for game in games])))
print("Average length of game where P1 uses NN is %f moves" % (np.mean([float(len(game)) for game in games2])))
print("Average length of game where P2 uses NN is %f moves" % (np.mean([float(len(game)) for game in games3])))
print("Average length of game where both use NN is %f moves" % (np.mean([float(len(game)) for game in games4])))

Average length of fully random game is 7.788300 moves
Average length of game where P1 uses NN is 6.762000 moves
Average length of game where P2 uses NN is 7.711000 moves
Average length of game where both use NN is 8.036000 moves


* As shown above, the games are a move shorter for player 1 and a big longer for player 2.

* Now, we play a game against our model and see how well it does. We let it make the first move.

In [25]:
# Create new board
board = initBoard()

# Move 1 (computer)
move = bestMove(board, model, 1)
board[move[0]][move[1]] = 1
printBoard(board)

 | | 
-----
 | | 
-----
 | |X


In [26]:
# Move 2 (human)
board[1][1] = 2
printBoard(board)

 | | 
-----
 |O| 
-----
 | |X


In [27]:
# Move 3 (computer)
move = bestMove(board, model, 1)
board[move[0]][move[1]] = 1
printBoard(board)

X| | 
-----
 |O| 
-----
 | |X


In [28]:
# Move 4 (human)
board[0][2] = 2
printBoard(board)

X| |O
-----
 |O| 
-----
 | |X


In [29]:
# Move 5 (computer)
move = bestMove(board, model, 1)
board[move[0]][move[1]] = 1
printBoard(board)

X| |O
-----
 |O| 
-----
X| |X


In [30]:
# Move 6 (human)
board[2][1] = 2
printBoard(board)

X| |O
-----
 |O| 
-----
X|O|X


In [31]:
# Move 7 (computer)
move = bestMove(board, model, 1)
board[move[0]][move[1]] = 1
printBoard(board)

X| |O
-----
X|O| 
-----
X|O|X


* Computer wins! This might be bad for humanity. :-)